bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -788,89 +810,392 @@ private long readLastLogId(File f) {
         }
-        if (null == channels) {
-            return;
+
+        private final FastThreadLocal<ByteBuf> sizeBufferForAdd = new FastThreadLocal<ByteBuf>() {
+            @Override
+            protected ByteBuf initialValue() throws Exception {
+                return Unpooled.buffer(4);
+            }
+        };
+
+        /*
+         * This method should be guarded by a lock, so callers of this method
+         * should be in the right scope of the lock.
+         */
+        @Override
+        public long addEntry(long ledger, ByteBuf entry, boolean rollLog) throws IOException {
+            int entrySize = entry.readableBytes() + 4; // Adding 4 bytes to prepend the size
+            BufferedLogChannel logChannel = getCurrentLogForLedgerForAddEntry(ledger, entrySize, rollLog);
+            ByteBuf sizeBuffer = sizeBufferForAdd.get();
+            sizeBuffer.clear();
+            sizeBuffer.writeInt(entry.readableBytes());
+            logChannel.write(sizeBuffer);
+
+            long pos = logChannel.position();
+            logChannel.write(entry);
+            logChannel.registerWrittenEntry(ledger, entrySize);
+
+            return (logChannel.getLogId() << 32L) | pos;
         }
-        Iterator<BufferedLogChannel> chIter = channels.iterator();
-        while (chIter.hasNext()) {
-            BufferedLogChannel channel = chIter.next();
-            try {
-                channel.flushAndForceWrite(false);
-            } catch (IOException ioe) {
-                // rescue from flush exception, add unflushed channels back
-                synchronized (this) {
-                    if (null == logChannelsToFlush) {
-                        logChannelsToFlush = channels;
-                    } else {
-                        logChannelsToFlush.addAll(0, channels);
+
+        boolean reachEntryLogLimit(BufferedLogChannel logChannel, long size) {
+            if (logChannel == null) {
+                return false;
+            }
+            return logChannel.position() + size > logSizeLimit;
+        }
+
+        boolean readEntryLogHardLimit(BufferedLogChannel logChannel, long size) {
+            if (logChannel == null) {
+                return false;
+            }
+            return logChannel.position() + size > Integer.MAX_VALUE;
+        }
+
+        abstract BufferedLogChannel getCurrentLogForLedger(long ledgerId);
+
+        abstract BufferedLogChannel getCurrentLogForLedgerForAddEntry(long ledgerId, int entrySize, boolean rollLog)
+                throws IOException;
+
+        abstract void setCurrentLogForLedgerAndAddToRotate(long ledgerId, BufferedLogChannel logChannel);
+
+        /*
+         * flush current logs.
+         */
+        abstract void flushCurrentLogs() throws IOException;
+
+        List<BufferedLogChannel> getRotatedLogChannels() {
+            return rotatedLogChannels;
+        }
+
+        @Override
+        public void flush() throws IOException {
```

Comment:
I am not convinced this change is correct. it seems to me this might only be needed when you changed from `volatile List<BufferedLogChannel>` to `final CopyOnWriteArrayList<BufferedLogChannel>`. 

for sake of no changing the behavior of single-log manager, I would prefer keeping the behavior as same as before.

Reply:
we missed discussing about this (order of flush in flush method)

i think it is incorrect to call flushrotatedlogs first and then flushcurrentlogs. Consider this scenario - if in between flushRotatedLogs, flushCurrentLogs calls current activelog is rotated and added to rotatedlogs list then we would miss flushing that log in this current flush call. So it has to be first flushCurrentLogs and then flushRotatedLogs.

even in the case of 'volatile List<BufferedLogChannel>' also i dont think it is correct. For the following reason - 

1) lets say flush method is called 
2) now as per the existing implementation if flushRotatedLogs is called first then lets after the execution of the following synchronized block (where 'logChannelsToFlush' is assigned to 'channels' and null is assigned to 'logChannelsToFlush'), lets say current active entrylog is rotated in another thread (because of reaching the size limit) then the current active entrylog will be added to new 'logChannelsToFlush' list. And when flushCurrentLogs is called it would flush the current active entrylog but not the one which is just rotated. So we missed flushing the just rotated log in this flush() call

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java#L816

Reply:
@reddycharan 

first, I don't think flush current log should happen before rotated logs. because the way how checkpoint works for single-log manager. if current synchronization logic has problem, we should fix the synchronization, rather than reversing the order.  I am not sure about per-ledger log manager. 

second, I don't think this can happen with current "volatile" list approach. because the list creation and swap is `synchronized`. if you can show me a sequence and reproduce the problem in a test, I am convinced this is a problem. otherwise I will prefer us keeping the logic unchanged.